<?xml version="1.0"?>
<launch>
  <arg name="use_debug" default="false" />
  <arg name="use_dr_camera_info" default="true" />
  <arg name="use_rqt" default="true" />
  <arg name="use_image" default="true" doc="publish an image from disk" />
  <arg name="image" default="$(find vimjay)/data/chess_640.png" />
  <arg name="image_topic" default="image" />
  <!-- can't do camera_info remapping, need to make sure the right camera_info is paired with the image topic -->
  <!-- arg name="camera_info_topic" default="camera_info" /-->
  <arg name="image_transport" default="raw" doc="subscribe to raw image, otherwise set 'compressed'" />

  <arg name="distort_then_rectify" default="true" />

  <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager"
      output="screen"/>

  <node name="image_pub" pkg="vimjay" type="image.py"
      if="$(arg use_image)"
      args="$(arg image)"
      output="screen" >
    <remap from="image" to="$(arg image_topic)" />
    <remap from="camera_info" to="camera_info_unused" />
    <param name="rate" value="10" />
  </node>

  <node name="dr_camera_info" pkg="vimjay" type="dr_camera_info.py"
      output="screen"
      if="$(arg use_dr_camera_info)" >
  </node>

  <!-- this is for rectifying then distorting -->
  <group ns="undistort" unless="$(arg distort_then_rectify)">
    <node pkg="nodelet" type="nodelet" name="load_rectify"
        args="load image_proc/rectify /nodelet_manager"
        output="screen">
      <remap from="image_mono" to="$(arg image_topic)" />
      <!-- this remapping doesn't work, because of image transport?  Tilde prefix doesn't work also -->
      <!--remap from="~camera_info" to="$(arg camera_info_topic)" /-->
      <remap from="image_rect" to="image_rect_color" />
      <param name="image_transport" value="$(arg image_transport)" />
    </node>
    <group ns="image_color" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>

    <group ns="image_mono" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>

    <group ns="image_rect" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>

    <group ns="image_rect_color" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>

    <group if="true" >
      <!--node name="distort" pkg="image_manip" type="distort_image"
        output="screen" -->
      <node pkg="nodelet" type="nodelet"
          name="load_distort0"
          args="load image_manip/Distort /nodelet_manager"
          output="screen">
        <remap from="image" to="image_rect_color" />
        <!-- remap from="camera_info" to="$(arg camera_info_topic)" /-->
      </node>
      <group ns="distorted/image" >
        <rosparam param="disable_pub_plugins">
          - 'image_transport/compressed'
          - 'image_transport/compressedDepth'
          - 'image_transport/ffmpeg'
          - 'image_transport/theora'
        </rosparam>
      </group>
    </group>
  </group>  <!-- distort then rectify -->

  <!-- this is for distorting then rectifying -->
  <group ns="distort" if="$(arg distort_then_rectify)">

    <!-- node name="distort" pkg="image_manip" type="distort_image"
      output="screen" -->
    <node pkg="nodelet" type="nodelet"
        name="load_distort1" output="screen"
        args="load image_manip/Distort /nodelet_manager" >
      <remap from="image" to="/image" />
      <!-- remap from="camera_info" to="$(arg camera_info_topic)" /-->
      <param name="use_debug" value="$(arg use_debug)" />
    </node>

    <node name="image_proc" pkg="image_proc" type="image_proc"
        if="true"
        output="screen" >
      <remap from="image_raw" to="distorted/image" />
      <remap from="camera_info" to="distorted/camera_info" />
    </node>

    <!-- visualize the difference between the input and rectified output -
        should be the same except for interpolation errors and where data is missing -->
    <!-- Need to fix this, not currently working -->
    <group if="false">
      <node name="gray_pub" pkg="vimjay" type="color"
          output="screen" >
        <!-- would be nice to derive these from CameraInfo input -->
        <param name="width" value="640" />
        <param name="height" value="480" />
        <remap from="image" to="gray" />
      </node>
      <group ns="gray_pub" >
        <rosparam param="disable_pub_plugins">
          - 'image_transport/compressed'
          - 'image_transport/compressedDepth'
          - 'image_transport/ffmpeg'
          - 'image_transport/theora'
        </rosparam>
      </group>

      <node name="diff_image" pkg="vimjay" type="iir_imagev"
          output="screen" >
        <rosparam param="b_coeffs">[1.0, 0.5, -0.5]</rosparam>
        <remap from="image_0" to="gray" />
        <remap from="image_1" to="/image" />
        <remap from="image_2" to="/distort/image_rect" />
        <remap from="filtered_image" to="diff_image" />
      </node>
      <group ns="diff_image" >
        <rosparam param="disable_pub_plugins">
          - 'image_transport/compressed'
          - 'image_transport/compressedDepth'
          - 'image_transport/ffmpeg'
          - 'image_transport/theora'
        </rosparam>
      </group>
    </group>  <!-- diff image for visualization -->

    <!-- http://answers.ros.org/question/255970/how-to-deactivate-image_transport-plugins/ -->
    <group ns="distorted/image" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>
    <group ns="image_color" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>
    <group ns="image_rect" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>
    <group ns="image_mono" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>
    <group ns="image_rect_color" >
      <rosparam param="disable_pub_plugins">
        - 'image_transport/compressed'
        - 'image_transport/compressedDepth'
        - 'image_transport/ffmpeg'
        - 'image_transport/theora'
      </rosparam>
    </group>
  </group>  <!-- if distort_then_rectify -->

  <node name="rqt_generic" pkg="rqt_gui" type="rqt_gui"
      args="--perspective-file $(find vimjay)/config/distort.perspective"
      output="screen"
      if="$(arg use_rqt)" >
  </node>
</launch>
